File: src/utils/astronomy.ts
import { rad, deg } from './navigationMath';

/**
 * 薄明時計算 (Twilight)
 * cos(H) = (sin(alt) - sin(l)sin(d)) / (cos(l)cos(d))
 *
 * lat: 緯度 (度)
 * dec: 赤緯 (度)
 * lon: 経度 (度、東正)
 * eqTime: 均時差 (分)
 * alt: 天体の高度角 (度)
 *   市民薄明: -6°
 *   航海薄明: -12°
 *   天文薄明: -18°
 *   日出没: -0.833° (大気差補正込み)
 */
export function twilightTime(
  lat: number, dec: number, lon: number, eqTime: number,
  alt: number,
) {
  const latR = rad(lat), decR = rad(dec), altR = rad(alt);
  const cosH = (Math.sin(altR) - Math.sin(latR) * Math.sin(decR)) /
    (Math.cos(latR) * Math.cos(decR));

  if (Math.abs(cosH) > 1) {
    // 現象なし（白夜または極夜）
    return { rise: NaN, set: NaN, possible: false };
  }

  const H = deg(Math.acos(cosH)); // 時角（度）
  const hHours = H / 15; // 時角（時間）

  // LMT of transit = 12h - eqTime/60
  const transit = 12 - eqTime / 60;
  const rise = transit - hHours - lon / 15;
  const set = transit + hHours - lon / 15;

  return { rise, set, possible: true, hourAngle: H };
}

/**
 * 全薄明時刻をまとめて計算
 */
export function allTwilights(
  lat: number, dec: number, lon: number, eqTime: number,
) {
  const sunrise = twilightTime(lat, dec, lon, eqTime, -0.833);
  const civil = twilightTime(lat, dec, lon, eqTime, -6);
  const nautical = twilightTime(lat, dec, lon, eqTime, -12);
  const astronomical = twilightTime(lat, dec, lon, eqTime, -18);

  return { sunrise, civil, nautical, astronomical };
}

/**
 * 索星: 恒星の高度と方位を計算
 * Sight Reduction: sin(h) = sin(l)sin(d) + cos(l)cos(d)cos(LHA)
 */
export function starAltAz(
  lat: number, dec: number, lha: number,
) {
  const latR = rad(lat), decR = rad(dec), lhaR = rad(lha);

  const sinH = Math.sin(latR) * Math.sin(decR) + Math.cos(latR) * Math.cos(decR) * Math.cos(lhaR);
  const alt = deg(Math.asin(Math.max(-1, Math.min(1, sinH))));

  const cosAz = (Math.sin(decR) - Math.sin(latR) * sinH) / (Math.cos(latR) * Math.cos(Math.asin(sinH)));
  let az = deg(Math.acos(Math.max(-1, Math.min(1, cosAz))));
  if (Math.sin(lhaR) > 0) az = 360 - az;

  return { alt, az };
}

File: NavCalc/NavCalc.Core/Calculations/Astronomy.cs
using System;
using static NavCalc.Core.Calculations.NavigationMath;

namespace NavCalc.Core.Calculations
{
    public static class Astronomy
    {
        /// <summary>
        /// 薄明時計算
        /// </summary>
        public static (double Rise, double Set, bool Possible, double HourAngle) TwilightTime(
            double lat, double dec, double lon, double eqTime, double alt)
        {
            double latR = Rad(lat), decR = Rad(dec), altR = Rad(alt);
            double cosH = (Math.Sin(altR) - Math.Sin(latR) * Math.Sin(decR))
                        / (Math.Cos(latR) * Math.Cos(decR));

            if (Math.Abs(cosH) > 1)
                return (double.NaN, double.NaN, false, 0);

            double h = Deg(Math.Acos(cosH));
            double hHours = h / 15.0;
            double transit = 12.0 - eqTime / 60.0;
            double rise = transit - hHours - lon / 15.0;
            double set = transit + hHours - lon / 15.0;

            return (rise, set, true, h);
        }

        /// <summary>
        /// 全薄明時刻をまとめて計算
        /// </summary>
        public static (
            (double Rise, double Set, bool Possible, double HourAngle) Sunrise,
            (double Rise, double Set, bool Possible, double HourAngle) Civil,
            (double Rise, double Set, bool Possible, double HourAngle) Nautical,
            (double Rise, double Set, bool Possible, double HourAngle) Astronomical
        ) AllTwilights(double lat, double dec, double lon, double eqTime)
        {
            var sunrise = TwilightTime(lat, dec, lon, eqTime, -0.833);
            var civil = TwilightTime(lat, dec, lon, eqTime, -6.0);
            var nautical = TwilightTime(lat, dec, lon, eqTime, -12.0);
            var astronomical = TwilightTime(lat, dec, lon, eqTime, -18.0);

            return (sunrise, civil, nautical, astronomical);
        }

        /// <summary>
        /// 索星: 恒星の高度と方位を計算
        /// </summary>
        public static (double Alt, double Az) StarAltAz(double lat, double dec, double lha)
        {
            double latR = Rad(lat), decR = Rad(dec), lhaR = Rad(lha);

            double sinH = Math.Sin(latR) * Math.Sin(decR) + Math.Cos(latR) * Math.Cos(decR) * Math.Cos(lhaR);
            double alt = Deg(Math.Asin(Math.Max(-1.0, Math.Min(1.0, sinH))));

            double cosAz = (Math.Sin(decR) - Math.Sin(latR) * sinH)
                         / (Math.Cos(latR) * Math.Cos(Math.Asin(sinH)));
            double az = Deg(Math.Acos(Math.Max(-1.0, Math.Min(1.0, cosAz))));
            if (Math.Sin(lhaR) > 0) az = 360.0 - az;

            return (alt, az);
        }
    }
}
